dot/state/epoch.go

```diff
@@ -63,80 +71,65 @@ type EpochState struct {
 	nextConfigDataLock sync.RWMutex
 	// nextConfigData follows the format map[epoch]map[block hash]next config data
 	nextConfigData nextEpochMap[types.NextConfigDataV1]
+
+	genesisEpochDescriptor *GenesisEpochDescriptor
 }
 
 // NewEpochStateFromGenesis returns a new EpochState given information for the first epoch, fetched from the runtime
 func NewEpochStateFromGenesis(db database.Database, blockState *BlockState,
 	genesisConfig *types.BabeConfiguration) (*EpochState, error) {
-	baseState := NewBaseState(db)
-
-	err := baseState.storeFirstSlot(1) // this may change once the first block is imported
-	if err != nil {
-		return nil, err
-	}
-
-	epochDB := database.NewTable(db, epochPrefix)
-	err = epochDB.Put(currentEpochKey, []byte{0, 0, 0, 0, 0, 0, 0, 0})
-	if err != nil {
-		return nil, err
-	}
-
 	if genesisConfig.EpochLength == 0 {
-		return nil, errors.New("epoch length is 0")
+		return nil, ErrEpochLengthCannotBeZero
 	}
 
 	s := &EpochState{
 		baseState:      NewBaseState(db),
 		blockState:     blockState,
-		db:             epochDB,
+		db:             database.NewTable(db, epochPrefix),
 		epochLength:    genesisConfig.EpochLength,
+		slotDuration:   genesisConfig.SlotDuration,
 		nextEpochData:  make(nextEpochMap[types.NextEpochData]),
 		nextConfigData: make(nextEpochMap[types.NextConfigDataV1]),
-	}
 
-	epochDataRaw := &types.EpochDataRaw{
-		Authorities: genesisConfig.GenesisAuthorities,
-		Randomness:  genesisConfig.Randomness,
+		genesisEpochDescriptor: &GenesisEpochDescriptor{
+			EpochData: &types.EpochDataRaw{
+				Authorities: genesisConfig.GenesisAuthorities,
+				Randomness:  genesisConfig.Randomness,
+			},
+			ConfigData: &types.ConfigData{
+				C1:             genesisConfig.C1,
+				C2:             genesisConfig.C2,
+				SecondarySlots: genesisConfig.SecondarySlots,
+			},
+		},
 	}
 
-	err = s.SetEpochDataRaw(0, epochDataRaw)
+	err := s.StoreCurrentEpoch(0)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("storing current epoch")
 	}
 
-	err = s.SetConfigData(0, &types.ConfigData{
-		C1:             genesisConfig.C1,
-		C2:             genesisConfig.C2,
-		SecondarySlots: genesisConfig.SecondarySlots,
-	})
+	err = s.setLatestConfigData(0)
 	if err != nil {
 		return nil, err
 	}
 
-	if err = s.baseState.storeEpochLength(genesisConfig.EpochLength); err != nil {
-		return nil, err
-	}
-
-	if err = s.baseState.storeSlotDuration(genesisConfig.SlotDuration); err != nil {
-		return nil, err
-	}
-
 	if err := s.baseState.storeSkipToEpoch(0); err != nil {
 		return nil, err
 	}
 
 	return s, nil
 }
 
+// TODO: remove NewEpochStateFromGenesis function
```

Comment:
is this still TODO?

Reply:
Not really, my idea was to have a single constructor but will require more changes, will leave as it is